### DISCUSSION TITLE
Highlight function declaration identifiers differently than other identifiers

### DISCUSSION THREAD

**Discussion by showbiz2k** [ionide/tree-sitter-fsharp#113] [Q&A] [answered]
Using Ionide in vs code, function declaration identifiers were highlighted differently than value declaration identifiers.  I found this especially helpful when making a value declaration and accidentally forgetting parameters thus performing partial function application and creating a function instead of a value.

Using neovim, with Ionide-vim plugin, I was inspecting the tree on some sample code it appears they can be differentiated, but I'm not sure what to make of that.  I'm new to neovim in general and not sure if I should be looking at my color scheme, configuring ionide, or configuring treesitter itself.

Any guidance would be greatly appreciated.

**Comment by vzarytovskii** [comment]
I think treesitter is not used in VSCode to colour F#. I think this discussion can be moved to https://github.com/ionide/ionide-vscode-fsharp cc @baronfel

**Reply by showbiz2k** [reply]
I was referring to experience of ionide in VSCode.  

In neovim, the function declaration identifiers are not highlighted in a different color than the value declaration identifiers.

**Reply by showbiz2k** [reply]
I've edited my question, hope that helps clarify...thanks for the response.

**Reply by vzarytovskii** [reply]
Oh, I see.

This repository is for tree-sitter, which is not used by default in neither vscode-ionide or ionide-vim. 

If you're explicitly using treesitter for F# in nvim, then it's correct place. If you're only using ionide-vim, then it's probably a question for their repo.

Sorry for confusing you.

**Comment by Nsidorenco** [comment]
Hi!

If I understood the question correctly you would like an identifier linked to function definition to be highlighted as a function throughout the document. 

In theory, it’s enable by the `locals.scm` query, but in neovim these queries are not used to do highlighting by default. 

Maybe there is a plugin somewhere that enables it, but out of the box the highlighting is limited to what can be defined in `highlight.scm`

**Reply by Eliemer** [reply, marked as answer]
distinguishing between functions and values in F# is near impossible without type information, which tree-sitter doesn't have (nor want). Ionide uses semantic tokening after typechecking a file to provide those more in-depth colors.
```fs
// value
let x = 1
// func as value
let y = fun () -> 1 // this is indistinguishable from just values based on syntax alone. same issue happens with partially applied functions, no way to know without type info.
// func declr
let z () = 1
```

**Reply by showbiz2k** [reply]
I see what you mean when inspecting the tree and looking at the difference between the y identifier and the z identifier.  Thanks for the help.
